api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java

```diff
@@ -49,54 +45,88 @@ protected ComplexData readComplexData(Obs obs, ValueComplex valueComplex, String
+		// will return a BufferedImage
+		// (since that is what ImageHandler getObs in Core returns)
+		obs = getParent().getObs(obs, AttachmentsConstants.IMAGE_HANDLER_VIEW);
+
+		int imageHeight = Integer.MAX_VALUE;
+		int imageWidth = Integer.MAX_VALUE;
 
 		// Get image dimensions
-		try {
-			BufferedImage image = ImageIO.read(savedFile);
-			imageHeight = image.getHeight();
-			imageWidth = image.getWidth();
-		} catch (IOException e) {
-			log.warn("The dimensions of image file '" + savedFileName
-					+ "' could not be determined, continuing with generating its thumbnail anyway.");
-		}
+		BufferedImage image = (BufferedImage) obs.getComplexData().getData();
+		imageHeight = image.getHeight();
+		imageWidth = image.getWidth();
+		saveThumbnailIfNeeded(obs, imageHeight, imageWidth);
 
-		try {
-			savedFileName = saveThumbnailOrRename(savedFile, imageHeight, imageWidth);
-		} catch (APIException e) {
-			getParent().purgeComplexData(obs);
-			throw new APIException("A thumbnail file could not be saved for obs with" + "OBS_ID='" + obs.getObsId()
-					+ "', " + "FILE='" + complexData.getTitle() + "'.", e);
-		}
+		// We invoke the parent to inherit from the file saving routines.
+		return new ValueComplex(complexData.getInstructions(), complexData.getMimeType(), obs.getValueComplex());
+	}
 
-		return new ValueComplex(complexData.getInstructions(), complexData.getMimeType(), savedFileName);
+	/**
+	 * <p>
+	 * If the image is over a certain dimension, it will create a small thumbnail
+	 * file alongside the original file to be used as thumbnail image.
+	 * </p>
+	 *
+	 * @param obs
+	 *            original obs
+	 * @param imageHeight
+	 *            image height
+	 * @param imageWidth
+	 *            image width
+	 * @return savedFileName new renamed file name or original file name
+	 */
+	public void saveThumbnailIfNeeded(Obs obs, int imageHeight, int imageWidth) {
+		if ((imageHeight <= THUMBNAIL_MAX_HEIGHT) && (imageWidth <= THUMBNAIL_MAX_WIDTH)) {
+			return;
+		} else {
+			String key = appendThumbnailSuffix(((ImageHandler) getParent()).parseDataKey(obs));
+			try {
+				storageService.saveData(outputStream -> {
+					Object data = obs.getComplexData().getData();
+					if (!(data instanceof BufferedImage)) {
+						throw new IllegalArgumentException(
+								"Expected a BufferedImage, but got " + data.getClass().getName());
+					}
+					BufferedImage image = (BufferedImage) data;
+					Thumbnails.of(image).size(THUMBNAIL_MAX_HEIGHT, THUMBNAIL_MAX_WIDTH)
+							.outputFormat(obs.getComplexData().getMimeType().split("/")[1].toLowerCase())
+							.toOutputStream(outputStream);
+				}, null, null, key);
+				// the above is a bit of hack... we pass in the entire value we want use as a key instead of specifying the filename,
+				// module ID and key suffix and having the storage service to generate the full key for us
+				// this is because we need to be able to recreate the key based on the key of the main image
```

Comment:
FYI... I can explain further if it's helpful.  In a future iteration, we might want to store the full thumbnail key as another "field" in the "valueComplex" but I didn't want to down that road at this point.